### PR TITLE
Completable result code throwable

### PR DESCRIPTION
**Pull request by jack-berg**
Related to #6306.

Discussed in the 3/28/24 Java SIG. Also need to adjust the behavior of HttpExporter and GrpcExporter (and potentially the related sender implementations) to fail CompletableResultCodes exceptionally. Planning on doing this in a followup PR, since there are details to work out about what the structure is with the failure exception.

### CODE REVIEW

**Comment by jarrodrobins** [comment]
Hi @jack-berg any updates on when this might be merged? I'd be happy to look into contributing with PRs to update the Http and Grpc exporters once this is merged (as discussed https://github.com/open-telemetry/opentelemetry-java/issues/6306#issuecomment-2128391093). Thanks.

**Comment by jack-berg** [comment]
@jarrodrobins - I'm blocked on reviews / approvals from @open-telemetry/java-approvers. I review from you is helpful as well, but won't be able to merge without an approval from an approver.

**Review by jkwatson — APPROVED** [review]
I'm not in love with the "getFailureThrowable" name, but I can't think of anything better that would be as clear. (I was thinking about "getFailureReason", but that seems too general).

🚢

**Review by jaydeluca — APPROVED** [review]

**Pull request merged**
